optics_functions/coupling.py

```diff
@@ -255,44 +271,111 @@ def closest_tune_approach(df: TfsDataFrame, qx: float = None, qy: float = None,
+    """ Closest tune approach calculated by Eq. (14) in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is like the persson estimate but divided by 1 + 4|F1001|^2 ."""
+    return _cta_persson(df, qx_frac, qy_frac) / (1 + 4 * df[F1001].abs() ** 2)
 
 
-def _cta_calaga(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+def _cta_hoydalsvik_alt(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+    """ Closest tune approach calculated by Eq. (14) without the s-term in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is like the persson_alt estimate but divided by 1 + 4|F1001|^2 ."""
+    return _cta_persson_alt(df, qx_frac, qy_frac) / (1 + 4 * df[F1001].abs() ** 2)
+
+
+def _cta_teapot(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
     """Closest tune approach calculated by Eq. (27) in [CalagaBetatronCoupling2005]_ .
     If F1010 is not given, it is assumed to be zero.
     """
-    f_diff = df["F1001"].abs() ** 2
-    with suppress(KeyError):
-        f_diff -= df["1010"].abs() ** 2
+    f_diff = df[F1001].abs() ** 2
+    with suppress(KeyError):  # this is the only estimate that uses sum resonance
+        f_diff -= df[F1010].abs() ** 2
 
     return (
         (np.cos(PI2 * qx_frac) - np.cos(PI2 * qy_frac))
         / (np.pi * (np.sin(PI2 * qx_frac) + np.sin(PI2 * qy_frac)))
         * (4 * np.sqrt(f_diff) / (1 + 4 * f_diff))
     )
+
+
+def _cta_teapot_franchi(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+    """Closest tune approach calculated by Eq. (12) in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is the teapot approach with the Franchi approximation. """
+    return 4 * (qx_frac - qy_frac) * df[F1001].abs() / (1 + 4 * df[F1001].abs() ** 2)
+
+
+def _get_weights_from_lengths(df: TfsDataFrame) -> Tuple[float, np.array]:
+    """Coefficients for the `persson` method. """
+    # approximate length of each element (ds in integral)
+    s_periodic = np.zeros(len(df) + 1)
+    s_periodic[1:] = df[S].to_numpy()
+    s_periodic[0] = df[S][-1] - df.headers[LENGTH]
+
+    # weight ds/(2*pi*R) * N (as we take the mean afterwards)
+    weights = np.diff(s_periodic) / df.headers[LENGTH] * len(df)
+    return weights
+
+
+def check_resonance_relation(df: DataFrame, to_nan: bool = False) -> DataFrame:
+    """Checks that F1001 >= F1010. If desired, sets the invalid points to NaN.
+    This is only used for checking in the
+    :func:`~optics_functions.coupling.closest_tune_approach` function,
+    but can be invoked by the user with ``to_nan = True`` and the resulting
+    DataFrame can then be passed to
+    :func:`~optics_functions.coupling.closest_tune_approach`
+
+    Args:
+        df (DataFrame): Dataframe containing the coupling columns.
+        to_nan (bool): If true, sets values where F1001 <= F1010 to NaN
+
+    Returns:
+        df, either as given or a copy with NaNs.
+    """
+    df = df.copy()
```

Comment:
Shouldn't we raise if F1001 or F1010 not in the dataframe columns?

Reply:
I feel like if this is the case and we just return we're misleading the user a bit, since it's definitely expected behavior to have those.

Reply:
as we assume |F1001| >> |F1010| anyway, I don't want to check for |F1010| and ignore if not given.

Reply:
Yeah this is why I'm wondering if we should simply raise.

Reply:
Ah, sorry, misunderstood where you want to raise. We don't raise, because it might be only at a very limited amount of places where the condition is not fulfilled. and then it's more or less fine. Hence the NaNs to ignore these values.
And also hence, the percentage in the warning, so the user can decide how much to trust the results.

Reply:
Yes I see, that's fair.